Clipboard: Fix some data races, add missing format checks

BenWiederhake opened the pull request:
In several places, parsing was done in multiple requests to ClipboardServer. Not only is this wasteful, it also means that the state can change between requests, which could cause data corruption or even crashes. To avoid issues like this in the future, I renamed the Clipboard methods to `fetch_data_and_type` and `fetch_mimetype`.

Furthermore, I found missing checks (Gfx::BitmapFormat for bitmaps, and bitmap size for glyphs) and even a typo (ClipboardHistory, which is [my fault](https://github.com/SerenityOS/serenity/commit/6c2ea4c4e9627fef0415a9e7fc50837a26e433fa)), and fixed those.

linusg commented on Userland/Libraries/LibGUI/Clipboard.cpp:
```diff
@@ -64,34 +64,34 @@ Clipboard::DataAndType Clipboard::data_and_type() const
         return nullptr;
 
-    auto width = clipping.metadata.get("width").value_or("0").to_uint();
+    auto width = this->metadata.get("width").value_or("0").to_uint();
```
Pretty sure these don't need `this->`?

linusg commented on Userland/Libraries/LibVT/TerminalWidget.cpp:
```diff
@@ -752,13 +752,12 @@ void TerminalWidget::paste()
+    if (data_and_type.data.is_empty())
         return;
-    send_non_user_input(text);
+    send_non_user_input(data_and_type.data);
```
could use structured bindings here to make this a bit nicer:

```cpp
auto [data, mime_type, _] = GUI::Clipboard::the().data_and_type();
if (!mime_type.starts_with("text/"))
    return;
if (data.is_empty())
    ...
```

probably applies elsewhere as well, I'm not going to point them all out individually.

BenWiederhake replied:
Wouldn't that copy everything? `auto&& [data, mime_type, _] = GUI::Clipboard::the().data_and_type();` is also permitted, I think that would move it instead.

linusg replied:
uh, guess you'd do `auto& [...]` then :)

BenWiederhake replied:
That wouldn't work, and I'm also wrong about needing `&&`. :D

linusg replied:
why not?

https://github.com/SerenityOS/serenity/blob/66c06e8efb7c5e2abbf81fdfccaa80f24a4edbe0/Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp#L362

BenWiederhake replied:
`GlyphEditorWidget.cpp:87:11: error: non-const lvalue reference to type 'GUI::Clipboard::DataAndType' cannot bind to a temporary of type 'GUI::Clipboard::DataAndType'`

As far as I understand, the returned DataAndType is a temporary, whereas `Optional::value()` returns an lvalue (`FunctionAndArgumentIndex&`).

linusg replied:
I see. Oh well, can't have all the nice things :^)

linusg commented on Userland/Libraries/LibGUI/Clipboard.cpp:
```diff
@@ -89,9 +89,17 @@ RefPtr<Gfx::Bitmap> Clipboard::DataAndType::as_bitmap() const
+    if (!Gfx::is_valid_bitmap_format(format.value()))
+        return nullptr;
+    auto bitmap_format = (Gfx::BitmapFormat)format.value();
+    // We cannot handle indexed bitmaps, as the palette would lost.
```
```suggestion
    // We cannot handle indexed bitmaps, as the palette would <missing a word> lost.
```

?

linusg reviewed (commented):
`LibGUI: Permit pasting empty text, avoid access to ClipboardService`

Enabling the action for empty text and then have nothing happen on paste seems like terrible UX to me, not sure I like that.

The "make more robust" commits could use some explanation of what exactly is being improved.

linusg commented on Userland/Applications/FontEditor/GlyphEditorWidget.cpp:
```diff
@@ -84,16 +84,16 @@ void GlyphEditorWidget::copy_glyph()
 void GlyphEditorWidget::paste_glyph()
 {
-    auto mime_type = GUI::Clipboard::the().mime_type();
+    auto [data, mime_type, metadata] = GUI::Clipboard::the().data_and_type();
```
:yakkiss:

BenWiederhake replied:
Indeed :D

BenWiederhake commented:
Changes since last push:
- Fixed typos
- Switched to structured bindings of the `struct DataAndType`
- Removed silly `this->` accesses
- Rebased

Empty text in the clipboard is irrelevant, for two reasons:
- Anything with text selection disables the copy action if the selection is empty
- `Clipboard::set_data()` with an empty buffer fails because `anon_create(0, …)` fails because `Process::sys$anon_create` returns EINVAL for that case.
So we will never have that case anyway; I'll adjust the commit message next.

I'll adjust the "`robust`" commit messages also.

linusg commented on Userland/Libraries/LibGUI/TextEditor.cpp:
```diff
@@ -1453,10 +1453,11 @@ void TextEditor::paste()
         return;
 
-    auto paste_text = Clipboard::the().data();
+    auto paste_text = data;
```
Is this copy needed?

BenWiederhake commented:
Changes since last push:
- Adjusted commit messages as suggested (thanks linusg! Also: Yes, those commit messages really are better now, but it's also surprisingly tedious to write them :sweat_smile:)
- Removed the silly copy (thanks linusg!)
- Fixed casing of "clipboard", "Clipboard", and removed non-existent "ClipboardService" from commit messages
